### PR TITLE
Serialization improvements

### PR DESCRIPTION
**Pull request by dtebbs**
- standard way to export variable assignments
- some remaining interface inconsistencies and composability fixes
- modified binary format with maximum compatibility with other implemenations. serializers for all relevant objects.

### CODE REVIEW

**Comment by dtebbs** [comment]
The build was failing (I think a cache got invalidated on the CI machine, and a commit of libff that was referenced by libff.js was pruned ... or something like that), so there were a bunch of errors in the zeth_contracts build.  Should be fixed by the final commit.

**Comment by AntoineRondelet** [comment]
@dtebbs - I see that you modified the submodule here. Was that necessary for the CI to pass?
I'm asking in the context of #361 for which the CI fails with:
```console
npm WARN The package scrypt is included as both a dev and production dependency.

npm ERR! code 1
npm ERR! Command failed: git submodule update -q --init --recursive
npm ERR! warning: templates not found in /tmp/pacote-git-template-tmp/git-clone-1caab2ca
npm ERR! fatal: remote error: upload-pack: not our ref 1710792e1b6da5905dadda19167475f458f2b2ba
npm ERR! Fetched in submodule path 'depends/libff', but it did not contain 1710792e1b6da5905dadda19167475f458f2b2ba. Direct fetching of that commit failed.
npm ERR! 

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2021-03-31T16_54_31_814Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! zeth-contracts@1.0.0 install: `ln -s ../../depends/ganache-cli node_modules/ganache-cli && cd node_modules/ganache-cli && npm install`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the zeth-contracts@1.0.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2021-03-31T16_54_33_344Z-debug.log
```
So I'm curious to know if you had the same here.

Also, I saw you requested a review here, so I changed the status of the PR from `Draft` to `Ready to review`. Please feel free to revert if my review was asked by mistake. Finally, I can see some unchecked boxes in your PR description, can I let you check them all if this PR is ready to go?

**Comment by dtebbs** [comment]
> @dtebbs - I see that you modified the submodule here. Was that necessary for the CI to pass?
> I'm asking in the context of #361 for which the CI fails with:
> ```
> So I'm curious to know if you had the same here.

Yes, that looks like the same issue.  While fixing the libff.js reference to a purged commit I hit other version related issues (stale version numbers in some package-lock.json files etc.).  I'll put just the submodule update in a new PR so we can merge it independently.  That should fix your PR too.

> Also, I saw you requested a review here, so I changed the status of the PR from `Draft` to `Ready to review`. Please feel free to revert if my review was asked by mistake. Finally, I can see some unchecked boxes in your PR description, can I let you check them all if this PR is ready to go?

Yes, no problem.  Thanks.

**Comment by AntoineRondelet** [comment]
> Yes, that looks like the same issue. While fixing the libff.js reference to a purged commit I hit other version related issues (stale version numbers in some package-lock.json files etc.). I'll put just the submodule update in a new PR so we can merge it independently. That should fix your PR too.

Great, thanks for confirming and moving the submodule update in another PR @dtebbs !

File: libzeth/core/group_element_utils.hpp [hunk 1]
```diff
@@ -12,18 +12,41 @@ namespace libzeth
...
+template<typename GroupT>
+void group_element_read_bytes(GroupT &point, std::istream &in_s);
+
+/// Write a collection of group elements as bytes to a stream, using
+/// group_element_write_bytes.
+template<typename GroupCollectionT>
+void group_elements_write_bytes(
+    const GroupCollectionT &points, std::ostream &out_s);
+
+/// Read a collection of group elements as bytes, usinng
```
Comment (AntoineRondelet) [inline libzeth/core/group_element_utils.hpp:46]: using

File: libzeth/core/group_element_utils.tcc [hunk 2]
```diff
@@ -73,21 +74,59 @@ void point_affine_read_json(GroupT &point, std::istream &in_s)
...
     GroupT result;
-    point_affine_read_json(result, ss);
+    group_element_read_json(result, ss);
     return result;
 }
 
+template<typename GroupT>
+void group_element_write_bytes(const GroupT &point, std::ostream &out_s)
+{
+    typename std::decay<GroupT>::type affine_p = point;
```
Comment (AntoineRondelet) [inline libzeth/core/group_element_utils.tcc:96]: Why do you need  the decay type of GroupT here?

File: libzeth/serialization/stream_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+#include <type_traits>
+
+namespace libzeth
+{
+
+/// Statically derive the type of the element contained in a (vector-like)
+/// collection.
+template<typename CollectionT>
+using MemberT =
+    typename std::decay<decltype((*(CollectionT *)nullptr)[0])>::type;
```
Comment (AntoineRondelet) [inline libzeth/serialization/stream_utils.hpp:20]: Same question as above here re: the decay types. Wouldn't it be sufficient to simply get the type of a member of the collection by using `decltype`? Why exactly do we need the surrounding `decay::type` here?

File: libzeth/snarks/groth16/groth16_snark.hpp [hunk 4]
```diff
@@ -28,45 +28,54 @@ template<typename ppT> class groth16_snark
...
+    /// Generate the proof (from the values set to the protoboard)
     static proof generate_proof(
         const libsnark::protoboard<libff::Fr<ppT>> &pb,
         const proving_key &proving_key);
 
+    /// Generate the proof (from given primary and auxiliary values)
+    static proof generate_proof(
+        const proving_key &proving_key,
+        const libsnark::r1cs_primary_input<libff::Fr<ppT>> &primary_input,
+        const libsnark::r1cs_auxiliary_input<libff::Fr<ppT>> auxiliary_input);
```
Comment (AntoineRondelet) [inline libzeth/snarks/groth16/groth16_snark.hpp:40]: The signature of this static member is slightly inconsistent with the one above (i.e. the `proving_key` is set as the first parameter, while it is the last one above). I'd recommend to keep "invariant parameters at the same place" for better consistency here, and thus modify
```cpp
    static proof generate_proof(
        const libsnark::protoboard<libff::Fr<ppT>> &pb,
        const proving_key &proving_key);
```
to
```cpp
    static proof generate_proof(
        const proving_key &proving_key
        const libsnark::protoboard<libff::Fr<ppT>> &pb);
```

File: libzeth/snarks/groth16/groth16_snark.tcc [hunk 5]
```diff
@@ -50,94 +56,138 @@ bool groth16_snark<ppT>::verify(
...
 template<typename ppT>
-std::ostream &groth16_snark<ppT>::keypair_write_bytes(
-    const typename groth16_snark<ppT>::keypair &keypair, std::ostream &out_s)
+void groth16_snark<ppT>::proof_write_json(
+    const typename groth16_snark<ppT>::proof &proof, std::ostream &out_s)
 {
-    proving_key_write_bytes(keypair.pk, out_s);
-    verification_key_write_bytes(keypair.vk, out_s);
-    return out_s;
+    out_s << "{\n  \"a\": " << group_element_to_json(proof.g_A)
```
Comment (AntoineRondelet) [inline libzeth/snarks/groth16/groth16_snark.tcc:154]: Why not use `\t` here?
i.e. `"{\n \"a\":` -> `"{\n\t\"a\":` etc.

File: libzeth/snarks/groth16/groth16_snark.hpp (hunk 4 above)
Comment (dtebbs) [reply libzeth/snarks/groth16/groth16_snark.hpp:40]: switched arg order

File: libzeth/core/group_element_utils.hpp (hunk 1 above)
Comment (dtebbs) [reply libzeth/core/group_element_utils.hpp]: fixed typo.  thanks.

File: libzeth/core/field_element_utils.hpp [hunk 6]
```diff
@@ -58,6 +58,17 @@ template<typename FieldT> std::string field_element_to_json(const FieldT &el);
 template<typename FieldT>
 FieldT field_element_from_json(const std::string &json);
 
+/// Write a field element as bytes. Base field elements are written in plain
+/// (non-Montgomery) form as fixed-size big-endian integers. Extension fields
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzeth/core/field_element_utils.hpp]: updated comment

File: verifier/verifier.cpp [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        po::variables_map vm;
+        po::store(parsed, vm);
+
+        if (0 == vm.count("vk_file")) {
+            throw po::error("vk_file not specified");
+        }
+        if (0 == vm.count("primary_input_file")) {
+            throw po::error("primary_input_file not specified");
+        }
+        if (0 == vm.count("proof_file")) {
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply verifier/verifier.cpp]: switched comparison order

File: libzeth/core/group_element_utils.tcc (hunk 2 above)
Comment (dtebbs) [reply libzeth/core/group_element_utils.tcc:96]: `GroupT` can sometimes include `const`, which prevents calls to `to_affine_coordinates`.  `std::decay<T>` strips away const, ptr, reference to expose the underlying type.

File: libzeth/serialization/stream_utils.hpp (hunk 3 above)
Comment (dtebbs) [reply libzeth/serialization/stream_utils.hpp:20]: The index operator returns a reference or const reference, whereas we need `MemberT` to be the type of the underlying object in the collection.  As above, `std::decay` strips away any const / ref parts of the type.

**Comment by dtebbs** [comment]
Thanks for the comments @AntoineRondelet . Hopefully all addressed now.  Note I added one small commit to end of this PR (to write out full assignments, required to generate test data for external provers), so you may want to review that in isolation.  Thanks.

File: libzeth/core/group_element_utils.tcc (hunk 2 above)
Comment (AntoineRondelet) [reply libzeth/core/group_element_utils.tcc:96]: Oh I see, thanks!

File: libzeth/snarks/groth16/groth16_snark.tcc (hunk 5 above)
Comment (AntoineRondelet) [reply libzeth/snarks/groth16/groth16_snark.tcc:154]: Alright, no big deal here. Let's keep it like that.

**Review by AntoineRondelet — APPROVED** [review]
LGTM overall - thanks @dtebbs 
The CI failed though, so I just restarted the jobs - let's see if that passes.

**Comment by dtebbs** [comment]
> LGTM overall - thanks @dtebbs
> The CI failed though, so I just restarted the jobs - let's see if that passes.

Ah, looks like I broke something in the final commit.  I'll check.

**Comment by dtebbs** [comment]
> > LGTM overall - thanks @dtebbs
> > The CI failed though, so I just restarted the jobs - let's see if that passes.
> 
> Ah, looks like I broke something in the final commit. I'll check.

@AntoineRondelet   Should be fixed now.  Sorry about that.

**Comment by AntoineRondelet** [comment]
Great, thanks @dtebbs - LGTM

**Pull request merged**